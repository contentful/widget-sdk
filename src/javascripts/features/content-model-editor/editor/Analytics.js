import * as Analytics from 'analytics/Analytics';
import { get } from 'lodash';

export const trackAddedField = (contentType, field) => {
  Analytics.track('content_modelling:field_added', {
    contentTypeId: contentType.sys.id,
    contentTypeName: contentType.name,
    fieldId: field.id,
    fieldName: field.name,
    fieldType: field.type,
    fieldItemType: get(field, 'items.type') || null,
    fieldLocalized: field.localized,
    fieldRequired: field.required,
  });
};

export const trackEnforcedButtonClick = (err) => {
  // If we get reason(s), that means an enforcement is present
  const reason = get(err, 'body.details.reasons', null);

  Analytics.track('entity_button:click', {
    entityType: 'contentType',
    enforced: Boolean(reason),
    reason,
  });
};
