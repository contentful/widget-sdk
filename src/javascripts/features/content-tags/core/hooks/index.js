export { useCreateTag } from 'features/content-tags/core/hooks/useCreateTag';
export { useUpdateTag } from 'features/content-tags/core/hooks/useUpdateTag';
export { useDeleteTag } from 'features/content-tags/core/hooks/useDeleteTag';
export { useReadTags } from 'features/content-tags/core/hooks/useReadTags';
export { useF36Modal } from 'features/content-tags/core/hooks/useF36Modal';
export { useScrollToTop } from 'features/content-tags/core/hooks/useScrollToTop';
export { useTagsRepo } from 'features/content-tags/core/hooks/useTagsRepo';
export { useCurrentSpaceFeature } from 'features/content-tags/core/hooks/useCurrentSpaceFeature';
export { useTagsFeatureEnabled } from 'features/content-tags/core/hooks/useTagsFeatureEnabled';
export { useSpaceContext } from 'features/content-tags/core/hooks/useSpaceContext';
export { useIsAdmin } from 'features/content-tags/core/hooks/useIsAdmin';
export { useIsInitialLoadingOfTags } from 'features/content-tags/core/hooks/useIsInitialLoadingOfTags';
