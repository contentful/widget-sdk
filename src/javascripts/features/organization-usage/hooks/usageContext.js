import React from 'react';
import { mapResponseToState } from '../services/UsageService';
import { periodToDates } from '../utils/periodToDates';
import { sum, get } from 'lodash';
import PropTypes from 'prop-types';

export const colours = ['#2E75D4', '#0EB87F', '#EA9005', '#8C53C2', '#CC3C52'];

const initialState = {
  isLoading: true,
  error: null,
  isAssetBandwidthTab: false,
  selectedMainTab: 'apiRequest',
  selectedPeriodIndex: 0,
  selectedSpacesTab: 'cma',
};

export const UsageStateContext = React.createContext();
export const UsageDispatchContext = React.createContext();

const reducer = (state, action) => {
  switch (action.type) {
    case 'SET_ORG_DATA': {
      const { periods } = action.value;
      const periodDates = periodToDates(periods[state.selectedPeriodIndex]);
      return {
        ...state,
        ...action.value,
        periodDates,
      };
    }
    case 'SET_USAGE_DATA': {
      const { periodicUsage } = mapResponseToState(action.value);
      const totalUsage = sum(periodicUsage.org.usage);
      return { ...state, periodicUsage, totalUsage };
    }
    case 'SET_ASSET_BANDWIDTH_DATA': {
      const assetBandwidthData = {
        usage: get(action.value, ['usage']),
        limit: get(action.value, ['limits', 'included']),
        uom: get(action.value, ['unitOfMeasure']),
      };
      return { ...state, assetBandwidthData };
    }
    case 'CHANGE_PERIOD': {
      return {
        ...state,
        selectedPeriodIndex: action.value,
        periodDates: periodToDates(state.periods[action.value]),
      };
    }
    case 'SET_LOADING': {
      return { ...state, isLoading: action.value };
    }
    case 'SWITCH_MAIN_TAB': {
      return {
        ...state,
        selectedMainTab: action.value,
        isAssetBandwidthTab: action.value === 'assetBandwidth',
      };
    }
    case 'SWITCH_SPACES_TAB': {
      return { ...state, selectedSpacesTab: action.value };
    }
    case 'SET_ERROR': {
      return { ...state, error: action.value };
    }
    default: {
      throw new Error(`Unhandled action type: ${action.type}`);
    }
  }
};

export const UsageProvider = ({ children, orgId }) => {
  const [state, dispatch] = React.useReducer(reducer, initialState);
  return (
    <UsageStateContext.Provider value={{ ...state, orgId }}>
      <UsageDispatchContext.Provider value={dispatch}>{children}</UsageDispatchContext.Provider>
    </UsageStateContext.Provider>
  );
};

UsageProvider.propTypes = {
  orgId: PropTypes.string,
};

export const useUsageState = () => {
  const context = React.useContext(UsageStateContext);
  if (!context) {
    throw new Error('useUsageState must be used within a UsageProvider');
  }
  return context;
};

export const useUsageDispatch = () => {
  const context = React.useContext(UsageDispatchContext);
  if (!context) {
    throw new Error('useUsageDispatch must be used within a UsageProvider');
  }
  return context;
};
