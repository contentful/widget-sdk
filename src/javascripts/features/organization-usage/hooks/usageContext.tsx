import React, { createContext, useReducer, Dispatch, Context, ReactNode } from 'react';
import { mapResponseToState } from '../services/UsageService';
import { periodToDates, PeriodToDatesArgs, PeriodToDatesResponse } from '../utils/periodToDates';
import { sum, get } from 'lodash';
import { noop } from 'lodash/fp';

export const colours = ['#2E75D4', '#0EB87F', '#EA9005', '#8C53C2', '#CC3C52'];

type UsageContextStateType = {
  isLoading: boolean;
  error?: Error | null | undefined;
  isAssetBandwidthTab: boolean;
  selectedMainTab: string;
  selectedPeriodIndex: number;
  selectedSpacesTab: string;
  isTeamOrEnterpriseCustomer?: boolean;
  periodDates?: PeriodToDatesResponse;
  periodicUsage?: {
    org: {
      usage: number[];
    };
    apis: any;
  };
  totalUsage?: number;
  apiRequestIncludedLimit?: number;
  assetBandwidthData?: {
    usage: any;
    limit: any;
    uom: any;
  };
  spaceNames: Record<string, string>;
  spaceTypeLookup: Record<string, any>;
  periods: PeriodToDatesArgs[];
  hasSpaces?: boolean;
  orgId?: string;
};

export const initialState = {
  isLoading: true,
  error: null,
  isAssetBandwidthTab: false,
  selectedMainTab: 'apiRequest',
  selectedPeriodIndex: 0,
  selectedSpacesTab: 'cma',
  spaceNames: {},
  spaceTypeLookup: {},
  periods: [] as PeriodToDatesArgs[],
};

export const UsageStateContext = createContext(initialState) as Context<UsageContextStateType>;
export const UsageDispatchContext = createContext(noop) as Context<Dispatch<any>>;

const reducer = (state, action) => {
  switch (action.type) {
    case 'SET_CUSTOMER_TYPE': {
      return { ...state, isTeamOrEnterpriseCustomer: action.value };
    }
    case 'SET_ORG_DATA': {
      const { periods } = action.value;
      const periodDates = periodToDates(periods[state.selectedPeriodIndex]);
      return {
        ...state,
        ...action.value,
        periodDates,
      };
    }
    case 'SET_USAGE_DATA': {
      const { periodicUsage } = mapResponseToState(action.value);
      const totalUsage = sum(periodicUsage.org.usage);
      return { ...state, periodicUsage, totalUsage };
    }
    case 'SET_ASSET_BANDWIDTH_DATA': {
      const assetBandwidthData = {
        usage: get(action.value, ['usage']),
        limit: get(action.value, ['limits', 'included']),
        uom: get(action.value, ['unitOfMeasure']),
      };
      return { ...state, assetBandwidthData };
    }
    case 'CHANGE_PERIOD': {
      return {
        ...state,
        selectedPeriodIndex: action.value,
        periodDates: periodToDates(state.periods[action.value]),
      };
    }
    case 'SET_LOADING': {
      return { ...state, isLoading: action.value };
    }
    case 'SWITCH_MAIN_TAB': {
      return {
        ...state,
        selectedMainTab: action.value,
        isAssetBandwidthTab: action.value === 'assetBandwidth',
      };
    }
    case 'SWITCH_SPACES_TAB': {
      return { ...state, selectedSpacesTab: action.value };
    }
    case 'SET_ERROR': {
      return { ...state, error: action.value };
    }
    default: {
      throw new Error(`Unhandled action type: ${action.type}`);
    }
  }
};

type UsageProviderProps = {
  children: ReactNode;
  orgId: string;
};

export const UsageProvider = ({ children, orgId }: UsageProviderProps) => {
  const [state, dispatch] = useReducer(reducer, initialState);
  return (
    <UsageStateContext.Provider value={{ ...state, orgId }}>
      <UsageDispatchContext.Provider value={dispatch}>{children}</UsageDispatchContext.Provider>
    </UsageStateContext.Provider>
  );
};

export const useUsageState = () => {
  const context = React.useContext(UsageStateContext);
  if (!context) {
    throw new Error('useUsageState must be used within a UsageProvider');
  }
  return context;
};

export const useUsageDispatch = () => {
  const context = React.useContext(UsageDispatchContext);
  if (!context) {
    throw new Error('useUsageDispatch must be used within a UsageProvider');
  }
  return context;
};
