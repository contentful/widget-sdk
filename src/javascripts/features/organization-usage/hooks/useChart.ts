import { useEffect, useState, useRef } from 'react';

import * as echarts from 'echarts';

export const useChart = (props) => {
  const chartRef = useRef<HTMLDivElement | null>(null);
  const [chart, setChart] = useState(null);

  useEffect(() => {
    if (chartRef.current) {
      setChart(echarts.init(chartRef.current));
    }
  }, []);

  useEffect(() => {
    if (chart) {
      // eslint-disable-next-line @typescript-eslint/ban-ts-ignore
      // @ts-ignore
      chart.setOption(props);
    }
  });

  return chartRef;
};
