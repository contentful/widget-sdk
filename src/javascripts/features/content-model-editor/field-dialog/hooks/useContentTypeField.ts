import { useCallback, useMemo, useState } from 'react';
import { getWidgetSettings } from '../utils/helpers';
import { useForm } from 'core/hooks';
import {
  getInitialValueFormFields,
  getNodeValidationsFormFields,
  getSettingsFormFields,
  getValidationsFormFields,
} from '../utils/formFieldDefinitions';
import { ContentType, ContentTypeField, EditorInterfaceControl } from 'core/typings';

type ContentTypeFieldProps = {
  contentType: ContentType;
  field: ContentTypeField;
  submitFn: Function;
  widget: EditorInterfaceControl;
};

function useWidgetSettings(widget) {
  const [data, setData] = useState(() => getWidgetSettings(widget));
  const [pristine, setPristine] = useState(true);
  const setActualData = useCallback((newData) => {
    setData(newData);
    setPristine(false);
  }, []);

  return {
    pristine,
    data,
    setData: setActualData,
  };
}

export function useContentTypeField({
  submitFn,
  contentType,
  field,
  widget,
}: ContentTypeFieldProps) {
  const widgetSettings = useWidgetSettings(widget);
  const form = useForm({
    fields: {
      ...getSettingsFormFields(field, contentType),
      ...getValidationsFormFields(field),
      ...getNodeValidationsFormFields(field),
      ...getInitialValueFormFields(field),
    },
    submitFn,
  });

  return useMemo(
    () => ({
      contentType: form,
      widgetSettings: {
        data: widgetSettings.data,
        setData: widgetSettings.setData,
      },
      get isInvalid() {
        return form.form.invalid;
      },
      get isPristine() {
        return form.form.pristine && widgetSettings.pristine;
      },
    }),
    [form, widgetSettings]
  );
}
