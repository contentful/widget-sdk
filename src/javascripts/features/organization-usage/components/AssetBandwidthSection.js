import React from 'react';
import { css } from 'emotion';
import { partialRight } from 'lodash';
import {
  TextLink,
  Note,
  Typography,
  Heading,
  Paragraph,
} from '@contentful/forma-36-react-components';
import * as tokens from '@contentful/forma-36-tokens';

import { track } from 'analytics/Analytics';
import { shortenStorageUnit } from 'utils/NumberUtils';
import { useUsageState } from '../hooks/usageContext';

const styles = {
  heading: css({
    color: '#536171',
    fontWeight: tokens.fontWeightNormal,
  }),
  usageNumber: css({
    color: '#253545',
    fontSize: tokens.fontSize3Xl,
    fontWeight: tokens.fontWeightMedium,
  }),
  limit: css({
    fontSize: tokens.fontSizeM,
  }),
  includedLimit: css({
    color: '#6A7889',
  }),
  overage: css({
    color: tokens.colorOrangeMid,
  }),
  learnMoreLink: css({
    marginLeft: tokens.spacingS,
  }),
  note: css({
    width: '40%',
    marginTop: tokens.spacingS,
    color: '#27457A',
  }),
};

export const AssetBandwidthSection = () => {
  const { assetBandwidthData } = useUsageState();

  const { limit, usage, uom } = assetBandwidthData;
  const withUnit = partialRight(shortenStorageUnit, uom);
  const handleClick = () => {
    track('usage:fair_use_policy_clicked');
  };
  return (
    <Typography>
      <Heading element="h2" className={styles.heading}>
        Total asset bandwidth
      </Heading>
      <Paragraph data-test-id="asset-bandwidth-usage" className={styles.usageNumber}>
        {withUnit(usage)}
      </Paragraph>
      <div className={styles.limit}>
        <strong data-test-id="asset-bandwidth-limit" className={styles.includedLimit}>{`${withUnit(
          limit
        )} included`}</strong>
        {usage > limit && (
          <small className={styles.overage} data-test-id="asset-bandwidth-overage">
            {' '}
            + {withUnit(usage - limit)} overage
          </small>
        )}
        <TextLink
          href="https://www.contentful.com/r/knowledgebase/fair-use/"
          target="_blank"
          onClick={handleClick}
          data-test-id="fair_use_policy_link"
          rel="noopener noreferrer"
          className={styles.learnMoreLink}>
          Fair Use Policy
        </TextLink>
        <Note className={styles.note}>
          Please note that asset bandwidth is calculated daily with a 48 hour delay.
        </Note>
      </div>
    </Typography>
  );
};
