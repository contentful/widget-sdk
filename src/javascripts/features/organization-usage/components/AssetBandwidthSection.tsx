import React from 'react';
import { css } from 'emotion';
import { partialRight } from 'lodash';
import {
  TextLink,
  Note,
  Typography,
  Heading,
  Paragraph,
  SkeletonContainer,
  SkeletonDisplayText,
  SkeletonBodyText,
} from '@contentful/forma-36-react-components';
import tokens from '@contentful/forma-36-tokens';

import { track } from 'analytics/Analytics';
import { shortenStorageUnit } from 'utils/NumberUtils';
import { useUsageState } from '../hooks/usageContext';

const styles = {
  heading: css({
    color: '#536171',
    fontWeight: tokens.fontWeightNormal,
  }),
  usageNumber: css({
    color: '#253545',
    fontSize: tokens.fontSize3Xl,
    fontWeight: tokens.fontWeightMedium,
  }),
  limit: css({
    fontSize: tokens.fontSizeM,
  }),
  includedLimit: css({
    color: '#6A7889',
  }),
  overage: css({
    color: tokens.colorOrangeMid,
  }),
  learnMoreLink: css({
    marginLeft: tokens.spacingS,
  }),
  note: css({
    width: '40%',
    marginTop: tokens.spacingS,
    color: '#27457A',
  }),
};

export const AssetBandwidthSection = () => {
  const { assetBandwidthData, isLoading } = useUsageState();

  const { limit, usage, uom } = assetBandwidthData ?? {};

  // Show loading state once at the initial data fetch
  const isPageLoading = isLoading && !assetBandwidthData;

  const withUnit = partialRight(shortenStorageUnit, uom);
  const handleClick = () => {
    track('usage:fair_use_policy_clicked', { source: 'Asset Bandwidth' });
  };

  const UsageAndLimit = () => (
    <>
      <Paragraph data-test-id="asset-bandwidth-usage" className={styles.usageNumber}>
        {
          // eslint-disable-next-line @typescript-eslint/ban-ts-comment
          // @ts-ignore
          withUnit(usage)
        }
      </Paragraph>
      <div className={styles.limit}>
        <strong data-test-id="asset-bandwidth-limit" className={styles.includedLimit}>{
          // eslint-disable-next-line @typescript-eslint/ban-ts-comment
          // @ts-ignore
          `${withUnit(limit)} included`
        }</strong>
        {usage > limit && (
          <small className={styles.overage} data-test-id="asset-bandwidth-overage">
            {' '}
            +{' '}
            {
              // eslint-disable-next-line @typescript-eslint/ban-ts-comment
              // @ts-ignore
              withUnit(usage - limit)
            }{' '}
            overage
          </small>
        )}
        <TextLink
          href="https://www.contentful.com/r/knowledgebase/fair-use/"
          target="_blank"
          onClick={handleClick}
          data-test-id="fair_use_policy_link"
          rel="noopener noreferrer"
          className={styles.learnMoreLink}>
          Fair Use Policy
        </TextLink>
      </div>
    </>
  );

  return (
    <Typography>
      <Heading element="h2" className={styles.heading}>
        Total asset bandwidth
      </Heading>
      {isPageLoading && (
        <SkeletonContainer svgWidth="40%" svgHeight="80px">
          <SkeletonDisplayText numberOfLines={1} />
          <SkeletonBodyText numberOfLines={1} offsetTop={50} />
        </SkeletonContainer>
      )}
      {!isPageLoading && <UsageAndLimit />}
      <Note className={styles.note}>
        Please note that asset bandwidth is calculated daily with a 48 hour delay.
      </Note>
    </Typography>
  );
};
