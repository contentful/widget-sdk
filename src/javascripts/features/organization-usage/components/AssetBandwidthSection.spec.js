import '@testing-library/dom';
import React from 'react';
import { render, fireEvent } from '@testing-library/react';
import { track } from 'analytics/Analytics';
import { AssetBandwidthSection } from './AssetBandwidthSection';
import { UsageStateContext } from '../hooks/usageContext';

const MockPovider = ({ children, assetBandwidthData }) => (
  <UsageStateContext.Provider value={{ assetBandwidthData }}>{children}</UsageStateContext.Provider>
);

const defaultData = {
  assetBandwidthData: {
    limit: 750,
    usage: 200,
    uom: 'GB',
  },
};

const renderComp = (data) => {
  return render(
    <MockPovider {...data}>
      <AssetBandwidthSection />
    </MockPovider>
  );
};

describe('AssetBandwidthSection', () => {
  it('should render', () => {
    const { container } = renderComp(defaultData);
    expect(container).toMatchSnapshot();
  });

  it('should render correct data', () => {
    const { getByTestId } = renderComp(defaultData);
    const bandwidthUsage = getByTestId('asset-bandwidth-usage');
    const bandwidthLimit = getByTestId('asset-bandwidth-limit');
    expect(bandwidthUsage.textContent).toBe('200 GB');
    expect(bandwidthLimit.textContent).toBe('750 GB included');
  });

  it('should render overage with correct data', () => {
    const overageData = {
      assetBandwidthData: {
        limit: 200,
        usage: 750,
        uom: 'GB',
      },
    };
    const { getByTestId } = renderComp(overageData);
    const overage = getByTestId('asset-bandwidth-overage');
    expect(overage.textContent).toBe(' + 550 GB overage');
  });

  it('should track fair use policy clicks', () => {
    const { getByTestId } = renderComp(defaultData);
    fireEvent.click(getByTestId('fair_use_policy_link'));
    expect(track).toHaveBeenCalledTimes(1);
    expect(track).toHaveBeenCalledWith('usage:fair_use_policy_clicked', {
      source: 'Asset Bandwidth',
    });
  });
});
