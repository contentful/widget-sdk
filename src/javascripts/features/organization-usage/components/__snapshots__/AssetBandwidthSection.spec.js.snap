// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AssetBandwidthSection should render 1`] = `
.emotion-0 {
  color: #536171;
  font-weight: 400;
}

.emotion-1 {
  color: #253545;
  font-size: 2.1875rem;
  font-weight: 600;
}

.emotion-4 {
  font-size: 0.875rem;
}

.emotion-2 {
  color: #6A7889;
}

.emotion-3 {
  margin-left: 0.75rem;
}

.emotion-5 {
  width: 40%;
  margin-top: 0.75rem;
  color: #27457A;
}

<div>
  <div
    class="Typography__Typography___1ZUfE"
    data-test-id="cf-ui-text-container"
  >
    <h2
      class="Heading__Heading___83D3K emotion-0 f36-margin-bottom--m"
      data-test-id="cf-ui-heading"
    >
      Total asset bandwidth
    </h2>
    <p
      class="Paragraph__Paragraph___2aO-9 emotion-1 f36-margin-bottom--m"
      data-test-id="asset-bandwidth-usage"
    >
      200 GB
    </p>
    <div
      class="emotion-4"
    >
      <strong
        class="emotion-2"
        data-test-id="asset-bandwidth-limit"
      >
        750 GB included
      </strong>
      <a
        class="TextLink__TextLink___1biUr a11y__focus-outline--default___2hwb1 emotion-3 TextLink__TextLink--primary___2Vc9F"
        data-test-id="fair_use_policy_link"
        href="https://www.contentful.com/r/knowledgebase/fair-use/"
        rel="noopener noreferrer"
        target="_blank"
      >
        <span
          class="TabFocusTrap__TabFocusTrap___39Vty"
          tabindex="-1"
        >
          Fair Use Policy
        </span>
      </a>
    </div>
    <div
      class="Note__Note___2eSKN emotion-5 Note__Note--primary___-8-7D"
      data-test-id="cf-ui-note"
    >
      <div
        class="Note__Note__icon___20RqC"
      >
        <svg
          class="Icon__Icon___38Epv Icon__Icon--small___1yGZK undefined"
          data-test-id="cf-ui-icon"
          height="18"
          viewBox="0 0 24 24"
          width="18"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M0 0h24v24H0z"
            fill="none"
          />
          <path
            d="M12 2C6.48 2 2 6.48 2 12s4.48 10 10 10 10-4.48 10-10S17.52 2 12 2zm1 15h-2v-6h2v6zm0-8h-2V7h2v2z"
          />
        </svg>
      </div>
      <div
        class="Note__Note__info___2m_-K"
      >
        <div>
          Please note that asset bandwidth is calculated daily with a 48 hour delay.
        </div>
      </div>
    </div>
  </div>
</div>
`;
